dialogue.go

```diff
@@ -190,26 +193,29 @@ func (d *Dialogue) MarshalLen() int {
 	if field := d.ObjectIdentifier; field != nil {
 		l += field.MarshalLen()
 	}
-	if field := d.SingleAsn1Type; field != nil {
-		l += field.MarshalLen()
-	}
 	if field := d.DialoguePDU; field != nil {
-		l += field.MarshalLen()
+		l += field.MarshalLen() + 2 // 2 = singleAsn1Type IE Header
 	}
-	l += len(d.Payload)
 
-	return l
+	return l + len(d.Payload)
 }
 
 // SetLength sets the length in Length field.
 func (d *Dialogue) SetLength() {
+	if field := d.ObjectIdentifier; field != nil {
+		d.ObjectIdentifier.SetLength()
+	}
+	if field := d.DialoguePDU; field != nil {
+		d.DialoguePDU.SetLength()
+	}
```

Comment:
No need to assign here

```suggestion
	if d.ObjectIdentifier != nil {
		d.ObjectIdentifier.SetLength()
	}
	if d.DialoguePDU != nil {
		d.DialoguePDU.SetLength()
	}
```